src/Entity/Form/ApiDocReimportSpecForm.php

```
@@ -0,0 +1,118 @@
+    parent::__construct($entity_repository, $entity_type_bundle_info, $time);
+    $this->messenger = $messenger;
+    $this->specFetcher = $spec_fetcher;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity.repository'),
+      $container->get('entity_type.bundle.info'),
+      $container->get('datetime.time'),
+      $container->get('messenger'),
+      $container->get('apigee_api_catalog.spec_fetcher')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getQuestion() {
+    return $this->t('Are you sure you want to update the OpenAPI specification file from URL on API Doc %name?', [
+      '%name' => $this->entity->label(),
+    ]);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCancelUrl() {
+    return new Url('entity.apidoc.collection');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDescription() {
+    return $this->t('This will replace the current OpenAPI specification file.
+     This action cannot be undone.');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    /* @var \Drupal\apigee_api_catalog\Entity\ApiDocInterface $entity */
+    $entity = $this->getEntity();
+
+    $needs_save = $this->specFetcher->fetchSpec($entity);
+    if ($needs_save) {
+      if ($entity->getEntityType()->isRevisionable()) {
+        $entity->setNewRevision();
+      }
+      $entity->save();
+      $this->messenger()->addStatus($this->t('API Doc %label: imported the OpenAPI specification file from URL.', [
+        '%label' => $this->entity->label(),
+      ]));
+    }
+    // TODO: $needs_save doesn't tell us if something went wrong.
```

This is exactly why I suggested earlier to throw an exception when file fetching or saving fails instead of returning a boolean indicator value which is not the OOP solution.

---

Yes, this can be addressed in a followup.

---

This TODO was added so it can be changed in a follow up task.

---

Which is the related follow-up task to this? Probably it would be better to create it as a GH issue, because that could be referenced in this PR.

---

https://github.com/apigee/apigee-api-catalog-drupal/issues/14